### PR TITLE
[Project] Add option to enrich project on load from project_setup.py file if it exists

### PR DESCRIPTION
**Pull request by yaronha**
Adding the option to run a setup script when loading a project 

In many cases there is a need to customize the project assets (metadata, functions, artifacts, feature sets, etc.), this PR enable that option:

* auto detects a `project_setup.py` and invoke the `setup()` handler in it
* pass the project object (loaded from the GIT, YAML or DB) as input and let the function modify/enrich the project (see example) 
* added the argument `parameters={}` to new/load_project()` (and in the CLI) so custom parameters can be used in the `setup` script (see example) 

Usage example (`project_setup.py` in the project context dir):

```python
def setup(project: mlrun.projects.MlrunProject):
    """Example for project setup script which modify project metadata and functions"""
    project.spec.params["test123"] = "456"

   # Create a job function and set an attribute (label)
    prep_func = project.set_function(
        "prep_data.py", "prep-data", kind="job", image="mlrun/mlrun"
    )
    prep_func.set_label("tst1", project.get_param("p2"))

    # create a serving function with a model
    srv_func = project.set_function(
        "serving.py", "serving", kind="serving", image="mlrun/mlrun"
    )
    # graph = srv_func.set_topology()
    srv_func.add_model("model", project.get_param("model_path"), class_name="MyCls")
    return project

```

### CODE REVIEW

**Review by theSaarco — COMMENTED** [review]
@yaronha, while this is a good addition and customization point, we need to be very careful not to make the `load_project` flow more confusing than it is today.
In context of this PR, the following are concerning me:
1. There's no doc whatsoever of this new option. This must be at least documented in the docstrings (and in MLRun docs, but it's reasonable for that to be in another PR)
2. The test added uses a zip file which I cannot review here, and is not easily visible to developers. Therefore the test is not a good reference point for how this handler needs to be constructed (what is the handler name, what the signature is etc.). I propose adding another test that will not use a zip file, but rather open files that can be easily viewed in the IDE
3. In general, since we're running user code here, it might be a good idea to wrap it in a try-catch block, and show a meaningful error to the user if something fails, like "custom project initialization handler code in project_setup.py failed". This will help someone loading a project received from somewhere else understand what failed if things go bad

File: tests/projects/test_project.py [hunk 1]
```diff
@@ -338,6 +338,21 @@ def test_load_project(
...
+        / "projects"
+        / "assets"
+        / "proj-setup.zip"
+    )
+    project = mlrun.load_project(context=context, url=url)
+    assert project.name == "projset"
+    assert project.spec.context == context
+    assert project.spec.source == str(url)
+    assert project.spec.params == {"label_column": "label", "test123": "456"}
+    print(project.to_yaml())
```
Comment (Tankilevitch) [inline tests/projects/test_project.py:353]: I would rather log the project before the assertion that way when debugging a failure you would be able to see the information

File: tests/projects/test_project.py [hunk 2]
```diff
@@ -338,6 +338,21 @@ def test_load_project(
...
 
 
+def test_load_project_with_setup(context):
+    url = (
+        pathlib.Path(tests.conftest.tests_root_directory)
+        / "projects"
+        / "assets"
+        / "proj-setup.zip"
+    )
+    project = mlrun.load_project(context=context, url=url)
```
Comment (Tankilevitch) [inline tests/projects/test_project.py:348]: I don't understand how this test your `project_setup.py` support

**Review by Tankilevitch — CHANGES_REQUESTED** [review]
It seems to miss user facing documentation on how to use it and what it is for.
also test doesn't look related.

File: tests/projects/test_project.py (hunk 1 above)
Comment (yaronha) [reply tests/projects/test_project.py:353]: added

File: tests/projects/test_project.py (hunk 2 above)
Comment (yaronha) [reply tests/projects/test_project.py:348]: placed the context as dir under assets (vs using zip) so you can see the project files and added explanations

**Comment by yaronha** [comment]
@theSaarco i added try and re-raise (since i want it to fail the load)

also moved it after the `.save()` so it would be possible to put things in the script which require the project to exist in the DB

**Review by theSaarco — APPROVED** [review]
Looks good, and test is now much clearer.
One thing on the docstring - it is on the private function `_run_project_setup`, where it won't be directly visible to developers and will not be visible in the MLRun docs. Can you move it to the user-facing functions, i.e. `load_project`, `get_or_create_project` etc.?

File: mlrun/projects/project.py [hunk 3]
```diff
@@ -427,14 +459,60 @@ def get_or_create_project(
...
+    if spec is None:
+        raise ImportError(f"cannot import project setup file in {setup_file_path}")
+    mod = imputil.module_from_spec(spec)
+    spec.loader.exec_module(mod)
+
+    if hasattr(mod, "setup"):
+        try:
+            project = getattr(mod, "setup")(project)
+        except Exception as exc:
+            logger.error(f"Failed to run project_setup script {setup_file_path}, {exc}")
```
Comment (liranbg) [inline mlrun/projects/project.py:503]:
```suggestion
            logger.error("Failed to run project_setup script", setup_file_path=setup_file_path, exc=err_to_str(exc))
```

File: tests/projects/assets/load_setup_test/prep_data.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+import mlrun
+
+
+def prep_data(context, source_url: mlrun.DataItem, label_column="label"):
+    # Convert the DataItem to a pandas DataFrame
+    df = source_url.as_df()
+    print("data url:", source_url.url)
+    df[label_column] = df[label_column].astype("category").cat.codes
+
+    # Record the DataFrane length after the run
```
Comment (liranbg) [inline tests/projects/assets/load_setup_test/prep_data.py:23]:
```suggestion
    # Record the DataFrame length after the run
```

File: mlrun/__main__.py [hunk 5]
```diff
@@ -1089,25 +1089,30 @@ def project(
...
         mlconf.dbpath = db
 
-    proj = load_project(context, url, name, init_git=init_git, clone=clone, save=save)
+    # set the CLI/GIT parameters in load_project() so they can be used by project setup scripts
+    parameters = fill_params(param) if param else {}
+    if git_repo:
+        parameters["git_repo"] = git_repo
+    if git_issue:
+        parameters["git_issue"] = git_issue
+    commit = environ.get("GITHUB_SHA") or environ.get("CI_COMMIT_SHA")
```
Comment (liranbg) [inline mlrun/__main__.py:1098]: what happened with `proj.get_param("commit_id")` precedence over envvars?

**Review by liranbg — CHANGES_REQUESTED** [review]
minor

File: mlrun/projects/project.py (hunk 3 above)
Comment (yaronha) [reply mlrun/projects/project.py]: done

File: tests/projects/assets/load_setup_test/prep_data.py (hunk 4 above)
Comment (yaronha) [reply tests/projects/assets/load_setup_test/prep_data.py]: done

File: mlrun/__main__.py (hunk 5 above)
Comment (yaronha) [reply mlrun/__main__.py:1098]: done

**Review by liranbg — APPROVED** [review]

**Pull request merged**